master/internal/resourcemanagers/priority.go

```diff
@@ -108,48 +115,53 @@ func (p *priorityScheduler) prioritySchedulerWithFilter(
 		successfulAllocations, unSuccessfulAllocations := trySchedulingPendingTasksInPriority(
 			allocationRequests, localAgentsState, fittingMethod)
 
-		// Only add these tasks to the lists of tasks to start if all tasks of higher priority
-		// have been scheduled.
-		if startTasks {
-			for _, allocatedTask := range successfulAllocations {
-				log.Debugf("scheduled task: %s", allocatedTask.Name)
+		// Only start tasks if there are no tasks of higher-priorities to preempt.
+		if len(toRelease) == 0 {
+			if !backfilling {
+				for _, allocatedTask := range successfulAllocations {
+					log.Debugf("scheduled task: %s", allocatedTask.Name)
+					toAllocate = append(toAllocate, allocatedTask)
+				}
+			} else if p.preemptionEnabled {
+				// Shiyuan: preemption and back-filling should always be enabled.
+				// A priority scheduler without preemption is useless in practice.
+				// We should remove the preemption configuration field.
+				for _, allocatedTask := range successfulAllocations {
+					if allocatedTask.NonPreemptible {
```

Comment:
It can't be merged because backfilling should only support preemptible tasks. If not, non-preemptible tasks would be backfilled and block large preemptible tasks to be scheduled.

Reply:
in the suggestion, wouldn't non-preemptible tasks be skipped too?

Reply:
> in the suggestion, wouldn't non-preemptible tasks be skipped too?

yes, they would be skipped in the backfilling.